xiilib/_gunicorn/wsgi_app.py

```diff
@@ -0,0 +1,131 @@
+
+"""Provide the WsgiApp class to represent the WSGI application."""
+
+import json
+import logging
+
+import ops
+
+from xiilib._gunicorn.charm_state import CharmState
+from xiilib._gunicorn.webserver import GunicornWebserver
+from xiilib.database_migration import DatabaseMigration
+
+logger = logging.getLogger(__name__)
+
+
+class WsgiApp:  # pylint: disable=too-few-public-methods
+    """WSGI application manager."""
+
+    def __init__(
+        self,
+        container: ops.Container,
+        charm_state: CharmState,
+        webserver: GunicornWebserver,
+        database_migration: DatabaseMigration,
+    ):
+        """Construct the WsgiApp instance.
+
+        Args:
+            container: The WSGI application container.
+            charm_state: The state of the charm.
+            webserver: The webserver manager object.
+            database_migration: The database migration manager object.
+        """
+        self._charm_state = charm_state
+        self._container = container
+        self._webserver = webserver
+        self._database_migration = database_migration
+
+    def _encode_env(self, value: str | int | float | bool | list | dict) -> str:
+        """Encode the environment variable values.
+
+        Args:
+            value: The input environment variable value.
+
+        Return:
+            The original string if the input is a string, or JSON encoded value.
+        """
+        return value if isinstance(value, str) else json.dumps(value)
+
+    def _wsgi_environment(self) -> dict[str, str]:
+        """Generate a WSGI environment dictionary from the charm WSGI configurations.
+
+        The WSGI environment generation follows these rules:
+            1. User-defined configuration cannot overwrite built-in WSGI configurations, even if
+                the built-in WSGI application configuration value is None (undefined).
+            2. Boolean and integer-typed configuration values will be JSON encoded before
+                being passed to application.
+            3. String-typed configuration values will be passed to the application as environment
+                variables directly.
+
+        Returns:
+            A dictionary representing the WSGI application environment variables.
+        """
+        config = self._charm_state.app_config
+        config.update(self._charm_state.wsgi_config)
+        prefix = f"{self._charm_state.framework.upper()}_"
+        env = {f"{prefix}{k.upper()}": self._encode_env(v) for k, v in config.items()}
+        secret_key_env = f"{prefix}SECRET_KEY"
+        if secret_key_env not in env:
+            env[secret_key_env] = self._charm_state.secret_key
+        for proxy_variable in ("http_proxy", "https_proxy", "no_proxy"):
+            proxy_value = getattr(self._charm_state.proxy, proxy_variable)
+            if proxy_value:
+                env[proxy_variable] = str(proxy_value)
+                env[proxy_variable.upper()] = str(proxy_value)
```

Comment:
Shouldn't this be managed from the CharmState directly ?

Reply:
Yes, it can. This is only used once, so it can put in any places.

Reply:
@jdkandersson WDYT ?

Reply:
If it can go anywhere then CharmState is probably the place to put it as a default